extensions/libevent.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright (C) 2020 Abdulrahman Semrie <hsamireh@gmail.com>
+ * Copyright (C) 2020 Aleix Conchillo Flaqué <aconchillo@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 3 of
+ * the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA
+ */
+
+
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+#include <event2/event.h>
+#include <libguile.h>
+
+struct event_data
+{
+  int fd;
+  short event;
+};
+
+struct wait_data
+{
+  int rv;
+  struct event_data *events;
+  int maxevents;
+};
+
+struct loop_data
+{
+  struct event_base *base;
+  int64_t timeout;
+};
+
+struct libevt_data
+{
+  struct event_base *base;
+};
+
+
+static void
+free_libevt (void *ptr)
+{
+  struct libevt_data *libevt = ptr;
+  event_base_free (libevt->base);
+  scm_gc_free(libevt, sizeof (struct libevt_data), "libevt_data");
+}
+
+static void
+free_wait_data (void *ptr)
+{
+  struct wait_data *data = ptr;
+  scm_gc_free(data, sizeof (struct wait_data), "wait_data");
+}
+
+static void
+cb_func (evutil_socket_t fd, short what, void *arg)
+{
+  struct wait_data* data = arg;
+  int rv = data->rv;
+
+  // In theory, we should never exceed the maximum number of events. This is
+  // because every time we add a new FD to monitor we check if we have enough
+  // room in our vector and if not we resize it. So, if this condition is true
+  // is because those assumptions are wrong and we might be doing something
+  // funky.
+  if (rv >= data->maxevents)
+    {
+      scm_puts ("fibers-libevent[ERROR]: ", scm_current_error_port ());
+      scm_puts ("max events fired on [", scm_current_error_port ());
+      scm_display (scm_from_int(fd), scm_current_error_port ());
+      scm_puts ("], ignoring.\n", scm_current_error_port ());
+      return;
```

Comment:
If some basic assumption is wrong, why ignore it, instead of, say, raise an exception?  That way, the test suites would test for it.

Also, if you go for not an exception, then please provide some option for applications to check whether this assumption has been violated -- I have in the past encountered some Fibers issues in non-Fibers test suites, I could insert a (when fibers-bad? (error "bad")) after the test suites.

Reply:
removed prints and calling `scm_syserror()` with a message.